rfcs/0013-module-versioning.md

```diff
@@ -0,0 +1,161 @@
+## Summary
+
+We describe ways to version modules containing types with serialized
+representations.
+
+## Motivation
+
+Within the Coda codebase, modules contain (ideally) one main type. As
+the software evolves, that type can change.  When compiling, the OCaml
+type discipline enforces consistent use of such types. Sometimes we'd
+like to export representations of those types, perhaps by persisting
+them, or by sending them to other software. Such other software
+may be written in languages other than OCaml.
+
+## Detailed design
+
+There are three main serialized representations of data used in the
+Coda codebase: `bin_io`, `sexp`, and `yojson`. The readers and writers
+of those representations are typically created using the `derive`
+annotation on types.
+
+### Explicit versioning
+
+The `bin_io` representation and associated `bin_prot` library are
+claimed to be type-safe when used in OCaml (see 
+[Jane Street bin\_prot](https://github.com/janestreet/bin_prot/).
+Moreover, other programming languages don't appear to be able to
+produce or consume this representation. Therefore, type and versioning
+information need not be stated explicitly in this
+representation. Nonetheless, it is still important to use versioning,
+so that a type of a particular name always has the same
+representation.
+
+Many programming languages can produce and consume the `sexp` and
+`yojson` representations. Given an OCaml type, the derived
+writers for those representation do not mention the type or
+a version. So non-OCaml consumers of these representations
+cannot easily detect the type distinctions in OCaml that gave
+rise to the representations.
+
+When a new version of a type is created, it can be worthwhile
+to retain the older version and associated code. That allows
+producing and consuming representations for older versions of the software.
+
+Suppose we create a module with a name and a version:
+
+```ocaml
+module Some_data = struct
+
+  let name = "some_data"
+
+  module Stable = struct
```

Comment:
Have you taken a look at `Async.Versioned_rpc` and made sure we do something that's easy to use with that?

Reply:
It looks like the query and response types need to have `bin_io` there, which is compatible with the regime here.

But again, I'm thinking of this module versioning as distinct from the RPC versioning, which has its own mechanisms.

Reply:
See my response to the earlier comment about RPC.